### PR TITLE
refacto netsurf parse document

### PR DESCRIPTION
**Pull request by krichprollsch**
relates to https://github.com/Browsercore/browsercore/pull/51#discussion_r1344264465

### CODE REVIEW

File: src/netsurf.zig [hunk 1]
```diff
@@ -553,11 +553,40 @@ fn documentHTMLVtable(doc_html: *DocumentHTML) c.dom_html_document_vtable {
...
 }
 
-pub fn documentHTMLParse(filename: []const u8) *DocumentHTML {
-    var f: []u8 = @constCast(filename);
-    const doc = c.wr_create_doc_dom_from_file(f.ptr);
+// documentHTMLParseFromFile reads the full document, loads the content in a
+// The allocator is required to create a null terminated string from filename.
+// The buffer is freed by the function.
+// The caller is responsible for closing the document.
+pub fn documentHTMLParseFromFile(allocator: std.mem.Allocator, filename: []const u8) !*DocumentHTML {
```
Comment (francisbouvier) [inline src/netsurf.zig]: We could remove the mandatory allocator by changing the filename type to `[:0]const u8`.

Maybe providing a `documentHTMLParseFromFileAlloc` helper.

File: src/netsurf.zig [hunk 2]
```diff
@@ -553,11 +553,40 @@ fn documentHTMLVtable(doc_html: *DocumentHTML) c.dom_html_document_vtable {
     return getVtable(c.dom_html_document_vtable, DocumentHTML, doc_html);
 }
 
-pub fn documentHTMLParse(filename: []const u8) *DocumentHTML {
-    var f: []u8 = @constCast(filename);
-    const doc = c.wr_create_doc_dom_from_file(f.ptr);
+// documentHTMLParseFromFile reads the full document, loads the content in a
```
Comment (francisbouvier) [inline src/netsurf.zig]: _in a_? `*DocumentHTML`?

File: src/netsurf.zig [hunk 3]
```diff
@@ -553,11 +553,40 @@ fn documentHTMLVtable(doc_html: *DocumentHTML) c.dom_html_document_vtable {
...
+
+    const doc = c.wr_create_doc_dom_from_file(cstr.ptr);
     if (doc == null) {
-        @panic("error parser");
+        return error.ParserError;
+    }
+    return @as(*DocumentHTML, @ptrCast(doc.?));
+}
+
+// documentHTMLParseFromCStrparses the given string.
```
Comment (francisbouvier) [inline src/netsurf.zig]: Space needed between _documentHTMLParseFromCStr_ and _parses_

File: src/netsurf.zig [hunk 4]
```diff
@@ -553,11 +553,40 @@ fn documentHTMLVtable(doc_html: *DocumentHTML) c.dom_html_document_vtable {
...
+// The allocator is required to create a null terminated string.
+// The c string allocated is freed by the function.
+// The caller is responsible for closing the document.
+pub fn documentHTMLParseFromStr(allocator: std.mem.Allocator, str: [:0]const u8) !*DocumentHTML {
```
Comment (francisbouvier) [inline src/netsurf.zig]: Here we should not required an allocator and the call to `dupeZ` has _str_ is already a `[:0]const u8`

File: src/netsurf.zig [hunk 5]
```diff
@@ -553,11 +553,40 @@ fn documentHTMLVtable(doc_html: *DocumentHTML) c.dom_html_document_vtable {
...
+    // create a null terminated c string.
+    const cstr = try allocator.dupeZ(u8, str);
+    defer allocator.free(cstr);
+
+    return documentHTMLParseFromCStr(cstr);
```
Comment (francisbouvier) [inline src/netsurf.zig]: I don't think `documentHTMLParseFromCStr` is required, we can directly call `c.wr_create_doc_dom_from_string(str.ptr)`

**Comment by krichprollsch** [comment]
I'm rebasing

File: src/netsurf.zig (hunk 5 above)
Comment (krichprollsch) [reply src/netsurf.zig]: Yes but having the zig signature is easier to understand and call IMO

Comment (krichprollsch) [reply src/netsurf.zig]: and we handle the `null` doc error also.

**Pull request merged**